src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -552,71 +528,76 @@ public Project WithName(string name, NameKind kind)
+
+            var resolvedProjectPath = Path.Combine(sourceDirectory, resolvedName.Replace('\\', '/'));
+            if (Directory.Exists(resolvedProjectPath))
             {
-                // If the path is a directory like '../lib', transform it to a project file path like '../lib/lib.csproj'.
-                // Also normalize backslashes to forward slashes to ensure the directive works on all platforms.
-                var sourceDirectory = Path.GetDirectoryName(sourceFile.Path)
-                    ?? throw new InvalidOperationException($"Source file path '{sourceFile.Path}' does not have a containing directory.");
-                var resolvedProjectPath = Path.Combine(sourceDirectory, resolvedName.Replace('\\', '/'));
-                if (Directory.Exists(resolvedProjectPath))
+                if (TryGetProjectFileFromDirectory(resolvedProjectPath, out var projectFilePath, out var error))
                 {
-                    var fullFilePath = GetProjectFileFromDirectory(resolvedProjectPath).FullName;
-
                     // Keep a relative path only if the original directive was a relative path.
                     resolvedName = ExternalHelpers.IsPathFullyQualified(resolvedName)
-                        ? fullFilePath
-                        : ExternalHelpers.GetRelativePath(relativeTo: sourceDirectory, fullFilePath);
+                        ? projectFilePath
+                        : ExternalHelpers.GetRelativePath(relativeTo: sourceDirectory, projectFilePath);
                 }
-                else if (!File.Exists(resolvedProjectPath))
+                else
                 {
-                    throw new GracefulException(FileBasedProgramsResources.CouldNotFindProjectOrDirectory, resolvedProjectPath);
+                    reportError(sourceFile, Info.Span, string.Format(FileBasedProgramsResources.InvalidProjectDirective, error));
                 }
             }
-            catch (GracefulException e)
+            else if (!File.Exists(resolvedProjectPath))
             {
-                diagnostics.AddError(sourceFile, Info.Span, string.Format(FileBasedProgramsResources.InvalidProjectDirective, e.Message), e);
+                reportError(sourceFile, Info.Span,
+                    string.Format(FileBasedProgramsResources.InvalidProjectDirective, string.Format(FileBasedProgramsResources.CouldNotFindProjectOrDirectory, resolvedProjectPath)));
             }
 
             return WithName(resolvedName, NameKind.ProjectFilePath);
         }
 
         // https://github.com/dotnet/sdk/issues/51487: Delete copies of methods from MsbuildProject and MSBuildUtilities from the source package, sharing the original method(s) under src/Cli instead.
-        private static FileInfo GetProjectFileFromDirectory(string projectDirectory)
+        private static bool TryGetProjectFileFromDirectory(string projectDirectory, [NotNullWhen(true)] out string? projectFilePath, [NotNullWhen(false)] out string? error)
```

Comment:
This method is a copy from MSBuildUtilities. Changing it only here will make it more difficult to share the code later as the comment above tracks. Consider either changing the original location too or sharing the code.